### PR TITLE
feat(svm-sdk): sealevel fee support warp deployments [IGNORE FOR NOW]

### PR DESCRIPTION
**Pull request by xeno097**
### Description

  Adds fee management support to SVM warp token deployments, allowing SVM warp tokens to configure on-chain fee programs like EVM already does.

  **Core changes:**
  - `SetFeeConfig` instruction builder (kind=8) for setting/removing fee config on warp tokens
  - Fee config reading from on-chain state via `splitPluginDataAndFeeConfig` decoder
  - Fee config diffing in the warp token update flow (add, remove, change program, change salt)
  - Program version detection via transaction simulation (`queryProgramVersion`)
  - Explicit program upgrade flow (`prepareProgramUpgrade`) with version comparison and downgrade rejection
  - Auto-appended `SetFeeConfig(None)` migration after upgrading from pre-fee binaries (required to realloc the token account for the new fee_config field)
  - `contractVersion` added to shared provider-sdk `BaseWarpConfig` types
  - `SvmDeployedWarpAddress` type extending `DeployedWarpAddress` with optional `feeConfig` for accurate diffing
  - Legacy program bytes preserved for E2E upgrade testing

  **Fee salt changes:**
  - Fee salt is now mandatory on all fee reader/writer constructors (no implicit defaults)
  - `resolveFeeSalt(chainName)` resolves salt from `SVM_FEE_{CHAIN_NAME}_SALT` env var, falling back to zeros
  - Artifact managers handle salt resolution transparently

  ### Drive-by changes

  - Replaced hand-rolled `compareVersions` with `compare-versions` library (same as EVM SDK) to correctly handle malformed/pre-release version strings
  - Added `compare-versions` to workspace catalog; migrated EVM SDK from pinned version to catalog reference
  - Unified duplicate `FeeConfigValue` and `TokenFeeConfig` into single `TokenFeeConfig` type
  - Added trailing bytes length validation in `splitPluginDataAndFeeConfig`
  - Removed unnecessary `as string` and `as Uint8Array` casts
  - Shared `PROGRAM_DATA_HEADER_SIZE` constant between `program-deployer.ts` and `warp-upgrade.ts`
  - Exported `SvmWarpArtifactManager` from index.ts (consistent with other VM SDKs)
  - Moved `deriveProgramDataAddress` from `program-deployer.ts` to `pda.ts`
  - Custom `getUpgradeInstruction`, `getExtendProgramCheckedInstruction`, and `getSetBufferAuthorityInstruction` using address-only signers (supports upgrade authority ≠ buffer writer)
  - Passed `contractVersion` through `validateWarpConfigForAltVM` for Alt-VM deploy path
  - Fixed double receipt collection in `prepareProgramUpgrade`
  - Fixed post-upgrade `SetFeeConfig(None)` migration to use token owner instead of upgrade authority

  ### Related issues

  - Full CLI compatibility for SVM fee deployment (tokenFee → provider-sdk FeeConfig mapping) will be implemented in a follow-up PR
  - Beneficiary ATA creation during fee program `setBeneficiary` will be addressed in a follow-up PR

  ### Backward compatibility

  No — intentional breaking changes:

  - Fee salt is now a required parameter on all SVM fee reader/writer constructors (previously defaulted to zeros). Public-facing artifact managers (`SvmWarpArtifactManager`, `SvmFeeArtifactManager`) resolve defaults via
  `resolveFeeSalt(chainName)`, so consumers going through the standard API are unaffected.
  - `SvmWarpArtifactManager` and `SvmFeeArtifactManager` constructors now require `chainConfig: { chainName: string }`.
  - `SvmDeployedWarpAddress` extends `DeployedWarpAddress` with optional `feeConfig` — readers now return this extended type.
  - `contractVersion` added to provider-sdk `BaseWarpConfig`, `BaseDerivedWarpConfig`, and `BaseWarpArtifactConfig` (optional field, no impact on other VMs).

  ### Testing

  **E2E tests (11 new tests):**
  - `warp-fee-config.e2e-test.ts` — set fee on deploy, add/change/remove fee via update, salt-change detection
  - `program-upgrade.e2e-test.ts` — deploy with legacy bytes → upgrade → read version, authority differs from payer, downgrade rejection, version match skip, migration required proof, migration success with ownership transfer

  **Unit tests (5 new tests):**
  - `warp-tx.unit-test.ts` — fee config diff: add, remove, change program, same program with different salt, no-op

  **Shared suite additions:**
  - `warp-token-suite.ts` — program version assertion after deploy, `supportsFeeConfig` validation

---

<a href="https://app.devin.ai/review/hyperlane-xyz/hyperlane-monorepo/pull/8676" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

### CODE REVIEW

**Review by paulbalaji — COMMENTED** [review]
## Review Summary

Review of SVM warp deploy fee support. Wire format (SetFeeConfig kind=8, account list, Borsh layout) lines up exactly with the Rust on-chain side from #8668; core diffing logic for add / remove / change-program is sound and well-structured.

### Existing review accounting

- **Devin's slot-boundary timing finding (`warp-upgrade.ts:187`) is current.** Inline below.
- **Devin's `SvmWarpArtifactManager` constructor positional break (`warp-artifact-manager.ts:41`) is current.** Inline below.

### High

- **Salt-only fee update can set `fee_config` on the token to an uninitialized fee_account PDA.** On-chain `set_fee_config` (#8668) accepts any (program, account) pair without verifying the PDA is initialized; subsequent `transferRemote` calls fail in fee CPI. Inline below.
- **Upgrade tx + migration tx emitted together without slot gating.** `DELAY_VISIBILITY_SLOT_OFFSET = 1` means the new binary is only callable in the slot *after* the Upgrade. Sequential `await signer.send()` typically takes more than one slot, but it isn't guaranteed. Inline below.
- **Half-upgraded recovery missing.** If the upgrade tx confirms but the migration tx fails (owner offline, RPC blip), `contractVersion` reads `1.0.0` while the token account is still legacy size. Next `update()` sees version match → migration never re-fires; admin instructions stay broken with `BorshIoError`. Inline below.

### Medium

- **`resolveFeeSalt` doesn't sanitize chain names with hyphens.** Inline below.
- **`SvmWarpArtifactManager` constructor positional change unflagged in changeset.** Inline below.
- **`queryProgramVersion` swallows transient RPC errors as null.** RPC hiccup on a 1.0.0 program → SDK treats as legacy → re-uploads binary + appends `SetFeeConfig(None)` migration that may overwrite operator-set fee config. Inline below.
- **Fee-diff unit tests assert tx count and annotation substring only.** Don't decode the instruction or verify it is `SetFeeConfig` (kind=8) with the correct Some/None tag and PDA — "remove fee" case in particular doesn't verify `null` is encoded; any other instruction with `'fee'` in the annotation would pass. Inline below.
- **Salt-only-change e2e test doesn't assert on-chain quote behavior** (`tests/warp-fee-config.e2e-test.ts:215`). Confirms token-side state changes; doesn't exercise transferRemote against the new PDA where the High issue above would surface.

### Low / nits / questions

- No unit test for `splitPluginDataAndFeeConfig`. Edge cases (zero-length, exact pluginSize, malformed Some tag = 2) only covered transitively via e2e.
- No e2e for negative auth — non-owner SetFeeConfig fails on-chain but SDK never exercises the path.
- Borsh trailing-byte comment in `version-query.ts:115` claims "always 0" but Rust sets `u8::MAX = 0xFF` (workaround for return-data trailing-zero truncation).
- `compareVersions` throws `TypeError` on malformed inputs — wrap with a labeled assert using `validate()`.
- Legacy program bytes (~5MB) ship in npm tarball via `dist/` — move under `src/testing/` or exclude e2e fixtures from build via `tsconfig` `exclude`.
- Custom loader instruction divergence risk vs `@solana-program/loader-v3` — add round-trip parity test or pin upstream version verified against.
- Migration `SetFeeConfig(None)` is unconditional — only safe today because legacy programs cannot have a fee_config; gate on `supportsFeeConfig(expectedVersion) && !supportsFeeConfig(currentVersion)`.
- `} as C` cast in `leaf-fee.ts:84` (still present from PR 8647; PR description claims removed all unnecessary casts).
- `DeployStageKind.Init` lookup uses `plan.stages` after `pop()` mutation — fragile.
- Combined-tx annotation always reads "ISM/hook/fee" even when only one of the three changed.
- Per-variant symmetry: native and synthetic destructure only `{ feeConfig }`; collateral and CC use both — inconsistent.
- No e2e exercises the slot-boundary race (just always sleeps).
- **Question**: is `contractVersion` consumed anywhere outside SVM? Added as optional on shared `BaseWarpConfig` types but currently SVM-only — should it be SVM-namespaced?
- **Operational**: are all already-deployed warp routes upgraded to a binary that includes `is_get_program_version` *before* this SDK is rolled out? Otherwise live programs return null and trigger spurious auto-upgrades.

Do not approve, do not request changes — leaving as comment for triage.

File: typescript/svm-sdk/src/warp/warp-tx.ts [hunk 1]
```diff
@@ -327,11 +363,49 @@ export async function computeWarpTokenUpdateInstructions(
...
+      !currentOnChainFeeConfig ||
+      !eqAddressSol(
+        currentOnChainFeeConfig.feeProgram,
+        expectedFeeConfig.feeProgram,
+      ) ||
+      !eqAddressSol(
+        currentOnChainFeeConfig.feeAccount,
+        expectedFeeConfig.feeAccount,
+      );
+    if (changed) {
```
Comment (paulbalaji) [inline typescript/svm-sdk/src/warp/warp-tx.ts:388]: **[High]** When the user changes only the fee salt (same fee program), this branch emits `SetFeeConfig` with the new `fee_account` PDA derived from the new salt. The on-chain `set_fee_config` (PR #8668, `lib/processor.rs:1149`) does **not** validate the fee_account exists or is initialized — the token will accept the new PDA, but subsequent `transferRemote` calls quoting fees against it will fail in the QuoteFee CPI because the new PDA has no fee account behind it.

Options: (a) reject salt-only changes in the diff and require operator to also redeploy the fee program at the new salt; (b) before emitting `SetFeeConfig`, fetch the fee account at `expectedFeePda.address` and `assert` it exists with the FeeAccount discriminator.

(See also salt-only-change e2e at `tests/warp-fee-config.e2e-test.ts:215` — that test only asserts the on-token state changes, not that the new PDA is usable. A `transferRemote` step after the salt change would have caught this.)

File: typescript/svm-sdk/src/warp/warp-upgrade.ts [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+  // that use store(account, false) — transfer_ownership, set_ism, set_igp —
+  // will fail with BorshIoError. SetFeeConfig(None) uses store_with_rent_exempt_realloc
+  // which grows the account and covers the rent delta.
+  //
+  // This MUST be a separate transaction — Solana's DELAY_VISIBILITY_SLOT_OFFSET
+  // means the new binary is only callable in the slot after the Upgrade tx.
+  // If this migration fails, the operator must manually send SetFeeConfig(None)
+  // to recover the program before any other config changes will succeed.
+  if (!supportsFeeConfig(currentVersion ?? null)) {
+    authorityTransactions.push({
```
Comment (paulbalaji) [inline typescript/svm-sdk/src/warp/warp-upgrade.ts:180]: **[High — confirms Devin's flag]** `prepareProgramUpgrade` returns the Upgrade tx and the `SetFeeConfig(None)` migration in a single `authorityTransactions` array. The doc comment above acknowledges the slot-visibility issue (`DELAY_VISIBILITY_SLOT_OFFSET = 1`) but provides no enforcement. Sequential `await signer.send()` typically takes more than one slot, but this is not guaranteed under fast RPC + LB conditions; same-slot dispatch will execute the migration against the old binary and fail with `InvalidInstructionData`.

Fix: capture the slot returned by the Upgrade `signer.send` receipt and submit the migration with `minContextSlot = upgradeSlot + 1` (or wait for `getSlot() > upgradeSlot`). Or split the return value (e.g. `{ upgrade: ..., migration: ... }`) so callers must explicitly sequence them.

**Companion finding (Medium):** if the upgrade tx confirms but the migration tx fails (owner offline, RPC blip, the same race above), `contractVersion` now reads `1.0.0` while the token account is still legacy-sized. On the next `update()`, line 62-64 sees `currentVersion === expectedVersion` and returns null — the auto-migration never fires again, and admin instructions all break with `BorshIoError`. The operator's only recovery is to hand-craft `SetFeeConfig(None)`. Detect the realloc-needed condition independently of version (e.g. fetch the token account, check `len < expected_v1_layout`) and re-emit the migration whenever the account size doesn't match; or expose a public `migrateRealloc(programId, owner)` helper for explicit recovery.

File: typescript/svm-sdk/src/fee/types.ts [hunk 3]
```diff
@@ -25,6 +25,25 @@ export function deriveFeeSalt(context: string): Uint8Array {
...
+ * Checks the SVM_FEE_{CHAIN_NAME}_SALT env var. If set, the value is
+ * hashed with keccak256 to produce a 32-byte salt. If not set, returns
+ * DEFAULT_FEE_SALT (all zeros).
+ *
+ * This ensures consistency across fee artifact managers and warp writers
+ * within the same deployment session without leaking protocol-specific
+ * details into the config.
+ */
+export function resolveFeeSalt(chainName: string): Uint8Array {
+  const envKey = `SVM_FEE_${chainName.toUpperCase()}_SALT`;
```
Comment (paulbalaji) [inline typescript/svm-sdk/src/fee/types.ts:40]: **[Medium]** `SVM_FEE_${chainName.toUpperCase()}_SALT` — chain names like `solana-mainnet` produce `SVM_FEE_SOLANA-MAINNET_SALT`. POSIX shells can't export environment variable names containing hyphens, so the var is silently ignored and the all-zeros default salt is used. Operators get no warning that their override didn't take effect.

Fix: replace `[^A-Z0-9_]` with `_` after upper-casing, or assert that `chainName` matches `/^[a-zA-Z0-9_]+$/` and fail loudly. Also worth adding a `logger.warn` when env-var lookup falls through to the zero default in any non-test environment.

File: typescript/svm-sdk/src/warp/warp-artifact-manager.ts [hunk 4]
```diff
@@ -31,6 +33,10 @@ import { detectWarpTokenType } from './warp-query.js';
 export class SvmWarpArtifactManager implements IRawWarpArtifactManager {
   constructor(
     private readonly rpc: Rpc<SolanaRpcApi>,
```
Comment (paulbalaji) [inline typescript/svm-sdk/src/warp/warp-artifact-manager.ts:35]: **[Medium — confirms Devin's flag]** `chainConfig` inserted at position 2; `ataPayerFundingAmount` shifted from position 2 to position 4. Both have defaults, so external callers `new SvmWarpArtifactManager(rpc, 200_000_000n)` will silently treat the bigint as `chainConfig` and crash later on `.chainName`.

The class is exported from `index.ts`. The PR's changeset (`svm-fee-sdk-support.md`) bumps `sealevel-sdk` as minor and documents a different breaking change (`createFeeArtifactManager`) but does not call out this positional shift.

Fix: convert constructor to a single options object — `{rpc, chainConfig, feeSalt?, ataPayerFundingAmount?}` — which is BC-safe for future additions and avoids positional foot-guns. Alternatively, bump sealevel-sdk to major and add the breakage to the existing changeset under an explicit "Breaking change" subheader.

File: typescript/svm-sdk/src/version/version-query.ts [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+      })
+      .send();
+
+    if (result.err) return null;
+
+    const returnData = result.returnData;
+    if (!returnData?.data?.[0]) return null;
+
+    const raw = Buffer.from(returnData.data[0], 'base64');
+    return decodeSimulationReturnDataString(raw);
```
Comment (paulbalaji) [inline typescript/svm-sdk/src/version/version-query.ts:102]: **[Medium]** `queryProgramVersion` returns `null` for both (a) program is pre-versioning legacy and (b) RPC simulation transient failure. A transient simulation error against an already-1.0.0 program causes the writer to treat it as legacy → `prepareProgramUpgrade` re-uploads the binary (3000+ writes) **and** appends a `SetFeeConfig(None)` migration that may overwrite a user-set fee config.

Fix: differentiate transient errors from "instruction not recognized" simulation results. Only return `null` when `result.err` matches `InvalidInstructionData` (legitimate "program doesn't recognize this discriminator"); rethrow / retry on network or parse failures so a hiccup can't trigger an unintended upgrade + clobber.

File: typescript/svm-sdk/src/tests/warp-tx.unit-test.ts [hunk 6]
```diff
@@ -108,3 +114,105 @@ describe('computeWarpTokenUpdateInstructions — feePayer', () => {
...
+      PROGRAM,
+      OWNER,
+      createStubRpc(),
+      'test',
+      DEFAULT_FEE_SALT,
+      undefined,
+    );
+
+    expect(txs).to.have.length(1);
+    expect(txs[0].annotation).to.include('fee');
```
Comment (paulbalaji) [inline typescript/svm-sdk/src/tests/warp-tx.unit-test.ts:147]: **[Medium]** All 5 fee-diff unit tests assert `expect(txs).to.have.length(1)` plus `expect(txs[0].annotation).to.include('fee')`. They do not decode `txs[0].instructions[0].data` to verify it is actually a `SetFeeConfig` (kind = 8) with the correct Some/None tag and PDA payload.

The "remove fee" case in particular (line 153 below) doesn't verify that `null` is encoded — any other instruction with `'fee'` in its annotation (or no fee instruction at all if the diff is wrong) would still pass.

Fix: decode via `decodeTokenProgramInstruction` (or whichever parser exists) and `assert` the instruction kind, plus assert the encoded value matches the expected `feeProgram` / `feeAccount` (or `null`).

**Comment by xeno097** [comment]
Superseded by https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/8717
